go/pools/resource_pool.go

```diff
@@ -188,6 +203,24 @@ func (rp *ResourcePool) closeIdleResources() {
 	}
 }
 
+// Do will Get a resource from the pool, and then call the provided function
+// with that resource. The resource returned by the provided function is Put
+// back into the pool; callers may use this to reopen resources by returning a
+// nil value for the resource. That resource is Put back regardless of
```

Comment:
I think the additional simplicity of `rp.Do(ctx, func)` might not be worth the risk of accidentally leaking locks if you have a function that returns `nil` on an error case. And `return nil, err` is such a pattern I would actually expect it to happen at some point.

If you don't have a specific use case in mind I'd say don't make the release optional -- you're really only saving 4 lines at that point which seems not a huge win.

Reply:
This is potentially highly specific to the ResourcePool API: if you call `Put(nil)` that has the effect of not re-using the existing resource and instead getting a new one via the pool's `Factory` vs if you `Put(x)` then `x` goes back into the pool for reuse. So IMO it's reasonable to have the func passed to `Do` to ensure it returns correctly with whether the resource it used should be discarded or not (regardless of error/non-error). But, either way, we're not leaking locks.

Reply:
I agree with Richard's comment here. I think is not worth the risk adding this extra function here.